Add "Increase contrast" checkboxes for "Light high contrast" and "Dark high contrast" that force high contrast in Automatic mode

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/work_packages/66396

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Allow users to assert theme contrast in automatic mode. When “Force high-contrast when in Light|Dark mode” is checked, the respective high-contrast themes are selected regardless of the OS contrast setting; otherwise, the user's OS preference is respected.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

<img width="1231" height="711" alt="Screenshot 2025-09-08 at 1 06 00 PM" src="https://github.com/user-attachments/assets/42e15ef3-020e-4fb8-ace3-325b75093300" />
<img width="1235" height="626" alt="Screenshot 2025-09-08 at 1 05 34 PM" src="https://github.com/user-attachments/assets/a6ef7895-5304-4174-8b35-4023aac3333e" />

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

Removes the "Light high contrast" and "Dark high contrast" options: Decouples the user theme preference data structure to store the color modes and contrast preferences separately. This allows the aforementioned fine-grained control.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
Add "Increase contrast" checkboxes for "Light high contrast" and "Dark high contrast" that force high contrast in Automatic mode

https://community.openproject.org/work_packages/66396

akabiru committed:
Use custom matchers and test helpers for better test dsl

akabiru committed:
Add user preferences data migration

Amends the data structure for the following theme preferences:
  - light_high_contrast → theme: "light" + increase_theme_contrast: true
  - dark_high_contrast → theme: "dark" + increase_theme_contrast: true

akabiru committed:
Fix logo contrast update logic

In light high contrast the "OpenProject" white logo is replaced with dark one.
All other themes rely just use the default.

akabiru committed:
Fix inverted visibility logic for theme contrast options

Correct the parameter naming and logic flow in toggleElementVisibility
to use shouldShow instead of shouldHide for better readability.
When sync with OS is enabled, auto contrast options are now properly
shown and single theme options are hidden.

akabiru committed:
Set conditional contast values in dom

akabiru committed:
Update the contrast mode captions to the correct text

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -38,14 +38,31 @@ class My::LookAndFeelForm < ApplicationForm
+
+    f.check_box_group(data: { my__look_and_feel_target: "autoThemeContrast" }) do |group|
+      group.check_box name: :enable_auto_light_theme_contrast,
+                      label: I18n.t("activerecord.attributes.user_preference.enable_auto_light_theme_contrast"),
```
```suggestion
                      label: UserPreference.human_attribute_name(:enable_auto_light_theme_contrast),
```

or if you override `model`:

```suggestion
                      label: attribute_name(:enable_auto_light_theme_contrast),
```

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -55,12 +72,7 @@ class My::LookAndFeelForm < ApplicationForm
-          label: theme[0]
-        )
-      end
+      comment_sort_order_options.each { |(label, value)| select.option(value:, label:) }
```
👍🏻

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -38,14 +38,31 @@ class My::LookAndFeelForm < ApplicationForm
+      theme_options_for_select.each { |(label, value)| select.option(value:, label:) }
+    end
+
+    f.check_box_group(data: { my__look_and_feel_target: "autoThemeContrast" }) do |group|
```
Primer Forms provide a `multi_input` component, which AFAICT is designed for exactly this use case. However the docs aren't great (in fact, [they're broken](https://primer.style/view-components/lookbook/pages/forms/inputs/multi_input/))

There is an example here:
- https://qa.openproject-edge.com/lookbook/inspect/primer/forms/multi_input_form

and source:
- https://github.com/opf/primer_view_components/blob/main/app/lib/primer/forms/dsl/multi_input.rb
- https://github.com/opf/primer_view_components/blob/main/app/lib/primer/forms/primer_multi_input.ts

It would be great if we could investigate using this.

akabiru replied:
Good shout, I'll have a look. I wasn't aware of this.

akabiru replied:
I gave this a try and I'm afraid it didn't work out. AFAICT, the logic assumes single HTMLInput elements which don't work well with the `autoThemeContrast` checkbox group. As the logic asserts, only a single input element is activated at a time.

Open to having another look with you if you feel I missed something obvious 🙂

myabc commented on app/models/user_preference.rb:
```diff
@@ -38,7 +38,8 @@ class UserPreference < ApplicationRecord
 
-  THEMES = %w[light light_high_contrast dark dark_high_contrast sync_with_os].freeze
+  COLOR_MODES = %w[light dark].freeze
+  THEMES = COLOR_MODES + %w[sync_with_os].freeze
```
I'd prefer to use symbols here. Also, `THEMES` itself isn't frozen.

```suggestion
  COLOR_MODES = %i[light dark].freeze
  THEMES = [*COLOR_MODES, :sync_with_os].freeze
```

myabc commented on config/locales/en.yml:
```diff
@@ -1285,6 +1283,13 @@ en:
+        increase_contrast: "Increase contrast"
+        increase_contrast_caption: "Enables high-contrast mode for the chosen colour mode."
+        enable_auto_light_theme_contrast: "Force high-contrast when in Light mode"
+        enable_auto_dark_theme_contrast: "Force high-contrast when in Dark mode"
```
I'm not sure why we use the more strongly worded "force high-contrast" when in automatic mode vs. "increase contrast" - perhaps more a question for @psatyal ?

myabc replied:
(Force high-contrast implies that we would still respect `prefers-contrast`)

akabiru replied:
> (Force high-contrast implies that we would still respect `prefers-contrast`)

I reckon that's what @psatyal had in mind, which makes sense. In which case, it is a miss on my part.

akabiru replied:
https://github.com/opf/openproject/pull/20206/commits/e6393b36819b72f00daef49396d30e69c6fbbecd

myabc commented on db/migrate/20250905204438_migrate_theme_preferences.rb:
```diff
@@ -0,0 +1,47 @@
+  def migrate_theme_preferences_to_new_structure
+    say "Migrate light_high_contrast -> light theme with increase_theme_contrast: true"
+    execute <<~SQL.squish
+      UPDATE user_preferences
+      SET settings = settings || '{"theme": "light", "increase_theme_contrast": true}'::jsonb
+      WHERE settings ->> 'theme' = 'light_high_contrast';
```
I guess we might want to consider creating a util function (e.g. `Migration::MigrationUtils::SettingRenamer.rename` if we end up having to do this often.

I find it weird that our `user_preferences` table uses jsonb but `Settings` is serialized to YAML! (which Postgres doesn't know how to deserialize)

akabiru replied:
> I guess we might want to consider creating a util function (e.g. `Migration::MigrationUtils::SettingRenamer.rename` if we end up having to do this often.
> 

I'm not sure we'll do this often to warrant a shared utility; it should be a one-off operation.

> I find it weird that our `user_preferences` table uses jsonb but `Settings` is serialized to YAML! (which Postgres doesn't know how to deserialize)

That's rather odd. I'm not sure what the reasons are for that— but I have confidence in the unit tests. In any case, I'll run a sanity check.

akabiru replied:
> 
> > I find it weird that our `user_preferences` table uses jsonb but `Settings` is serialized to YAML! (which Postgres doesn't know how to deserialize)
> 
> That's rather odd. I'm not sure what the reasons are for that— but I have confidence in the unit tests. In any case, I'll run a sanity check.

AFAICS, settings are serialized to `HashWithIndifferentAccess` https://github.com/opf/openproject/blob/70f8ae898932bbec31bf7f071824c92e0dc62fdd/app/models/user_preference.rb#L34 I did not find the YAML serialization. 👀

myabc replied:
Sorry, that was just an observation - no need to address anything.

myabc commented on frontend/src/stimulus/controllers/dynamic/my/look-and-feel.controller.ts:
```diff
@@ -0,0 +1,63 @@
+  declare singleThemeContrastTarget:HTMLElement;
+
+  private syncWithOsValue = 'sync_with_os';
+  private displayNoneClass = 'd-none';
```
Probably not necessary, but Stimulus supports configurable classes: https://stimulus.hotwired.dev/reference/css-classes

We are very inconsistent in how we hide things in our codebase, using a combination of `class="d-none"`, `style="display:none"` and the newer `hidden` attribute. I would probably prefer the latter, since it also takes care of accesibility, but it would be good to reach some consensus on this - perhaps when @HDinger is back.

akabiru replied:
Configurable classes seemed like overkill in this instance. The `hidden` attribute is a good call, and I agree we should have a consensus on this. 👍🏾

myabc commented on frontend/src/stimulus/controllers/dynamic/my/look-and-feel.controller.ts:
```diff
@@ -0,0 +1,63 @@
+
+  private toggleElementVisibility(targetElement:HTMLElement, shouldShow:boolean) {
+    const container = targetElement.closest(this.viewComponentSelector);
+    container?.classList.toggle(this.displayNoneClass, !shouldShow);
```
If we're hiding with a CSS class (e.g. `d-none`), then we should also set `aria-hidden` for good measure. / @bsatarnejad

myabc commented on spec/features/my/support/theme_matchers.rb:
```diff
@@ -0,0 +1,91 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+RSpec::Matchers.define :have_theme do |theme, options = {}|
+  match do |page|
+    contrast_suffix = options[:high_contrast] ? "_high_contrast" : ""
+    page.has_css?("body[data-color-mode='#{theme}'][data-#{theme}-theme='#{theme}#{contrast_suffix}']")
+  end
+
+  failure_message do |_page|
+    contrast_text = options[:high_contrast] ? " with high contrast" : ""
+    "expected page to have #{theme} theme#{contrast_text}"
+  end
+
+  failure_message_when_negated do |_page|
+    contrast_text = options[:high_contrast] ? " with high contrast" : ""
+    "expected page not to have #{theme} theme#{contrast_text}"
+  end
+end
```
Bit of an nitpick (as ultimately, they achieve the same thing), but a custom Capybara selector would probably be better suited here.

See `spec/support/capybara/primer_component_selectors.rb` for some examples.

Something like this:

```
Capybara.add_selector :theme do
  xpath do |locator|
    xpath = XPath["body"]
    xpath = builder(xpath).add_attribute_conditions("data-color-mode": locator)
    xpath
  end
end
```

akabiru replied:
Sure, I’d be happy to give it a try. 🙂 I usually avoid XPath syntax because I don’t find it intuitive, but this seems straightforward enough!

myabc replied:
You can use CSS selectors as well - see `:primer_label` selector.

akabiru replied:
https://github.com/opf/openproject/pull/20206/commits/de3232b64ae5a3c75c39895173f57d70710fdfdb

myabc reviewed (changes requested):
Code-wise looks very good, although there are a few things (e.g. `MultiInput`) that would worth looking into before we hit merge.

In terms of behaviour however, while I agree that end-users should be able to override the default OS contrast setting, I think it's a shame (and even a regression) that we no longer respect `prefers-contrast` by default in auto-contrast mode.

akabiru commented:
> Code-wise looks very good, although there are a few things (e.g. `MultiInput`) that would worth looking into before we hit merge.
> 

https://github.com/opf/openproject/pull/20206#discussion_r2344448133

> In terms of behaviour however, while I agree that end-users should be able to override the default OS contrast setting, I think it's a shame (and even a regression) that we no longer respect `prefers-contrast` by default in auto-contrast mode.

https://github.com/opf/openproject/pull/20206#discussion_r2344442727 True, this should be reinstated! 👍🏾

akabiru committed:
Respect OS contrast preference when "force" setting is false

When automatic mode is selected and the "force" checkboxes are not checked. Honor,
the user's OS contrast setting

akabiru committed:
Use `AR#human_attribute_name` for more elegant translations

akabiru committed:
Prefer symbols and freeze constants

akabiru committed:
Use `hidden` attribute for better a11ly

akabiru committed:
Add custom Capybara selectors for theme testing

Replaces raw CSS selectors with reusable Capybara selectors following
the existing pattern used in primer_component_selectors.rb.

- Creates :theme selector for testing color modes and contrast settings
- Creates :auto_theme_config selector for auto theme configuration
- Updates interface tests to use new selectors with cleaner syntax
- Consolidates theme test helpers inline in interface_spec.rb
- Removes duplicate theme_matchers.rb file

akabiru committed:
Optimize colorModeContrastPreferences initialization

akabiru committed:
Use 'force' terminology for theme contrast preferences

Changes attribute names from enable_auto_*_contrast to force_*_contrast
to better match existing UI terminology and differentiate from single
theme contrast settings.

akabiru commented on app/models/user_preference.rb:
```diff
@@ -38,7 +38,8 @@ class UserPreference < ApplicationRecord
   WORKDAYS_FROM_MONDAY_TO_FRIDAY = [1, 2, 3, 4, 5].freeze
 
-  THEMES = %w[light light_high_contrast dark dark_high_contrast sync_with_os].freeze
+  COLOR_MODES = %i[light dark].freeze
+  THEMES = (COLOR_MODES + %i[sync_with_os]).freeze
```
No. It is also used to define `base_theme_light?` and `base_theme_dark?` dynamic methods. https://github.com/opf/openproject/pull/20206/files#diff-c36f041a59833f85660511c3f91bab34e164096d161cba7083804c4d1a5926b9R148

akabiru commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -34,42 +34,68 @@ class My::LookAndFeelForm < ApplicationForm
+  end
+
+  def disable_keyboard_shortcuts_caption
+    attribute_name(:disable_keyboard_shortcuts_caption_html,
+                   href: OpenProject::Static::Links.url_for(:shortcuts)).html_safe # rubocop:disable Rails/OutputSafety
```
Static links are not user generated. They are predefined static URLs.

myabc commented:
@akabiru I'm getting an error locally. Is one of the translations missing?

<img width="1050" height="528" alt="image" src="https://github.com/user-attachments/assets/7e53e707-801e-49fc-8ecc-340d4e5e050b" />

akabiru commented:
> @akabiru I'm getting an error locally. Is one of the translations missing?
> <img alt="image" width="1050" height="528" src="https://private-user-images.githubusercontent.com/755/489697079-7e53e707-801e-49fc-8ecc-340d4e5e050b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgwMTM5NzcsIm5iZiI6MTc1ODAxMzY3NywicGF0aCI6Ii83NTUvNDg5Njk3MDc5LTdlNTNlNzA3LTgwMWUtNDlmYy04ZWNjLTM0MGQ0ZTVlMDUwYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkxNlQwOTA3NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zOGUwM2NkMmEzZjEzODc4MGI1YjVjMGFjM2NjNDFhMDc1ZDM3NThkYzcwMGE1NjY4ZjMxYzNkZTIyNWZjNTRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.a6NuBheVTd2Nr8Y9Sh20PVxeU4yPkAZ6NjQ6K0PbywI">

@myabc this is due to the rename of preferences from `enable_auto_{dark|light}_theme_contrast` to `force_*_contrast` https://github.com/opf/openproject/pull/20206/commits/0ac66975e664be584b21d8771597e60040650bb5 your local db still has the old traces.

You can remove them manually as follows: This won't be a problem for anyone who hasn't ran this PR locally before.

```
$ bin/rails dbconsole
```

```psql
openproject@[local]:5432 openproject_qa_local# BEGIN;
BEGIN
openproject@[local]:5432* openproject_qa_local# UPDATE user_preferences SET settings = (settings - 'enable_auto_dark_theme_contrast') WHERE settings ->> 'theme' = 'sync_with_os';
UPDATE 3
openproject@[local]:5432* openproject_qa_local# UPDATE user_preferences SET settings = (settings - 'enable_auto_light_theme_contrast') WHERE settings ->> 'theme' = 'sync_with_os';
UPDATE 3
openproject@[local]:5432* openproject_qa_local# COMMIT;
```

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -34,42 +34,68 @@ class My::LookAndFeelForm < ApplicationForm
+
+  def user_preference
+    User.current.pref
+  end
```
This method is never called, so can be removed.

akabiru replied:
It is used in the form definitions. E.g. `user_preference.increase_theme_contrast?`

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -34,42 +34,68 @@ class My::LookAndFeelForm < ApplicationForm
+
+  def attribute_name(...)
+    UserPreference.human_attribute_name(...)
+  end
```
The method signature could be changed in `ApplicationForm`.

akabiru replied:
Good shout 👍🏾

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -34,42 +34,68 @@ class My::LookAndFeelForm < ApplicationForm
+    f.check_box_group(data: { my__look_and_feel_target: "autoThemeContrast" }) do |group|
+      group.check_box name: :force_light_theme_contrast,
+                      label: attribute_name(:force_light_theme_contrast),
+                      checked: user_preference.force_light_theme_contrast?,
```
`checked` is redundant here.

```suggestion
```

akabiru replied:
Why is it redundant? Does primer form implicitly read the defined `name` property?

myabc replied:
Yes, Primer Forms are just a wrapper around Rails `FormBuilder`, which will call `force_light_theme_contrast` method on the model.

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -34,42 +34,68 @@ class My::LookAndFeelForm < ApplicationForm
+                      caption: attribute_name(:force_light_theme_contrast_caption)
+      group.check_box name: :force_dark_theme_contrast,
+                      label: attribute_name(:force_dark_theme_contrast),
+                      checked: user_preference.force_dark_theme_contrast?,
```
```suggestion
```

myabc commented on app/forms/my/look_and_feel_form.rb:
```diff
@@ -34,42 +34,68 @@ class My::LookAndFeelForm < ApplicationForm
+    f.check_box_group(data: { my__look_and_feel_target: "singleThemeContrast" }) do |group|
+      group.check_box name: :increase_theme_contrast,
+                      label: attribute_name(:increase_contrast),
+                      checked: user_preference.increase_theme_contrast?,
```
```suggestion
```

myabc reviewed (approved):
Thanks for looking into `multi_input`!

This looks good and works well. A couple minor comments relating to `LookAndFeelForm`, but should be mergeable!

akabiru committed:
Update `ApplicationForm#attribute_name` method signature to be less restrictive

Just forward all arguments to ActiveRecord's #huma_attribute_name. Esp when there's interpolation needed.

akabiru committed:
Remove seeming redundant check_box checked|value definitions.

Automatically inferred by primer forms
